### ISSUE TITLE
Unexpected ref safety error reported when a collection expression is used to construct a ref struct with Add method

### ISSUE THREAD

**Issue by jjonescz** [dotnet/roslyn#76374]
**Version Used**: main 2024-12-10

**Steps to Reproduce**:

```cs
using System.Collections;
class C
{
    R M()
    {
        var local = 1;
        return [local]; // error CS9203: A collection expression of type 'R' cannot be used in this context because it may be exposed outside of the current scope.
    }
}
ref struct R : IEnumerable
{
    public void Add(in int x) { }
    IEnumerator IEnumerable.GetEnumerator() => throw null;
}
```

**Expected Behavior**: Compiles successfully, the `in int x` parameter of the `Add` method cannot be captured into the ref struct. Replacing the collection expression with a collection initializer `new() { local };` also works fine. So does making the struct a non-ref struct. See also https://github.com/dotnet/roslyn/pull/76237#discussion_r1878652403.

**Actual Behavior**: Error is reported: "CS9203: A collection expression of type 'R' cannot be used in this context because it may be exposed outside of the current scope."

**Comment by RikkiGibson** [comment]
https://github.com/dotnet/csharplang/blob/main/proposals/csharp-12.0/collection-expressions.md#ref-safety

It looks like the spec supports what Jan is saying, it is defined to have *safe-context* *caller-context*.

It looks like a similar error occurs when `Add` signature is `public void Add(int x) { }`. [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgPgAQAwAIEEYB0BhA9gG3wFMBjAFylxgGcBuAWACgEAmFbJgbyZV5QCUUAWQAUASh59ujPrJQA3CACcU+XCQj4UAXhToGMubwQB2FAG01G/AF1aKAPQOURJUtwrsAZQCcLJADMICgAgigkBMTklDAuAB4ADkpE1NQxKLgAZihkAJ4JRCgA5PxF4RAwMLhkKMCFAK7URAAmKLA5ABZQ1OFUZERxNXUajYVQNZC5tYUDCbhNrbj1ZGnNhVmdhST1bkQwNdQRBZiSvAC+TBeMydnUZEr15AIowRgBXKdoAWgALKHNzREsBqcTEKE4KCusgAkgBRGD1MCuCBkDxodABTAAcSIZHhiORqKU4h0AD5Ou4AO4oBGEAxnIA==).
